fe2-android/app/src/main/java/com/github/dedis/popstellar/model/network/method/message/MessageGeneral.java

```diff
@@ -27,90 +26,63 @@
 
   private final String TAG = MessageGeneral.class.getSimpleName();
 
-  private final byte[] sender;
-
-  private final byte[] dataBuf;
-
+  private final PublicKey sender;
+  private final Base64URLData dataBuf;
   private final Data data;
+  private final MessageID messageId;
 
-  private byte[] signature;
-
-  private byte[] messageId;
-
+  private Signature signature;
   private List<PublicKeySignaturePair> witnessSignatures = new ArrayList<>();
 
-  private final PublicKeyVerify verifier;
-
-  public MessageGeneral(byte[] sender, Data data, PublicKeySign signer, Gson gson) {
-    this.sender = sender;
-    this.data = data;
-    Log.d(TAG, gson.toJson(data, Data.class));
-    this.dataBuf = gson.toJson(data, Data.class).getBytes();
-    this.verifier = new Ed25519Verify(sender);
-
-    generateSignature(signer);
-    generateId();
-  }
-
-  public MessageGeneral(
-      byte[] sender,
-      Data data,
-      List<PublicKeySignaturePair> witnessSignatures,
-      PublicKeySign signer,
-      Gson gson) {
-    this(sender, data, signer, gson);
-    this.witnessSignatures = witnessSignatures;
-  }
-
   public MessageGeneral(
-      byte[] sender,
-      byte[] dataBuf,
+      PublicKey sender,
+      Base64URLData dataBuf,
       Data data,
-      byte[] signature,
-      byte[] messageId,
+      Signature signature,
+      MessageID messageID,
       List<PublicKeySignaturePair> witnessSignatures) {
-    byte[] decodedMessageId = Base64.getUrlDecoder().decode(messageId);
-    Log.d(
-        TAG,
-        "new MessageGeneral with messageId encoded as: "
-            + new String(messageId, StandardCharsets.UTF_8)
-            + " decoded as: "
-            + Hex.bytesToStringUppercase(decodedMessageId));
     this.sender = sender;
-    this.messageId = messageId;
     this.dataBuf = dataBuf;
+    this.data = data;
+    this.messageId = messageID;
     this.signature = signature;
     this.witnessSignatures = witnessSignatures;
+  }
+
+  public MessageGeneral(KeyPair keyPair, Data data, Gson gson) {
+    this.sender = keyPair.getPublicKey();
     this.data = data;
-    this.verifier = new Ed25519Verify(sender);
+    Log.d(TAG, gson.toJson(data, Data.class));
```

Comment:
I tried to keep my already big PR as small as possible, and I did not apply too much changes.

It would probably be 2000 lines otherwise.

But I can change that, it makes a lot of sense.

Reply:
Done